pkg/vm/vm.go

```diff
@@ -321,19 +321,19 @@ func (v *VM) LoadScriptWithFlags(b []byte, f callflag.CallFlag) {
 // accordingly). It's up to the user of this function to make sure the script and hash match
 // each other.
 func (v *VM) LoadScriptWithHash(b []byte, hash util.Uint160, f callflag.CallFlag) {
-	v.loadScriptWithCallingHash(b, nil, v.GetCurrentScriptHash(), hash, f, 1, 0)
+	v.loadScriptWithCallingHash(b, nil, v.GetCurrentScriptHash(), hash, f, 1, 0, nil)
 }
 
 // LoadNEFMethod allows to create a context to execute a method from the NEF
 // file with the specified caller and executing hash, call flags, return value,
 // method and _initialize offsets.
 func (v *VM) LoadNEFMethod(exe *nef.File, caller util.Uint160, hash util.Uint160, f callflag.CallFlag,
-	hasReturn bool, methodOff int, initOff int) {
+	hasReturn bool, methodOff int, initOff int, onContextUnload ContextUnloadCallback) {
```

Comment:
This `onContextUnload` can handle isolation-related things as well, I think. `contract.Call` will set everything up and this callback will commit/discard and change interop context accordingly.

Reply:
I did this in a separate commit. That was tricky and I broke the logic while moving isolation-related code out of VM. I hope that currently it works correct, but the last commits need to be carefully reviewed. I'm sure that all commits before the https://github.com/nspcc-dev/neo-go/pull/2508/commits/3243d09ac07055bd9035da887414c8fe010ede90 implemented correctly and work as intended.

Reply:
Also, there's one thing that I was not able to remove from the VM context. It's notifications counter. The problem is that when discarding changes on exception, we need to know the exact amount of notifications to be reverted. This information can be properly handled only by VM context, because interop context doesn't have layered notifications structure.

Reply:
And I don't like the last commit, the code becomes very "spagettish" and non-clear. Maybe we still can keep an old version and just drop the last commit? Or improve it somehow, but I'm not sure, how.